relayer/cmd/subscribe_beefy.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		panic(err)
+	}
+	defer sub.Unsubscribe()
+
+	for {
+		select {
+		case msg := <-ch:
+
+			signedCommitment := &store.SignedCommitment{}
+			err := types.DecodeFromHexString(msg.(string), signedCommitment)
+			if err != nil {
+				log.WithError(err).Error("Failed to decode BEEFY commitment messages")
+			}
+
+			blockNumber := signedCommitment.Commitment.BlockNumber
+
+			log.WithField("commitmentBlockNumber", blockNumber).Info("Witnessed a new BEEFY commitment: \n")
+			if len(signedCommitment.Signatures) == 0 {
+				log.Info("BEEFY commitment has no signatures, skipping...")
+				continue
+			}
+			log.WithField("blockNumber", blockNumber+1).Info("Getting hash for next block")
+			nextBlockHash, err := relaychainConn.GetAPI().RPC.Chain.GetBlockHash(uint64(blockNumber + 1))
+			if err != nil {
+				log.WithError(err).Error("Failed to get block hash")
+			}
+			log.WithField("blockHash", nextBlockHash.Hex()).Info("Got blockhash")
+			GetMMRLeafForBlock(uint64(blockNumber), nextBlockHash, relaychainConn)
+			header, err := GetParaheads(nextBlockHash, relaychainConn)
+			if err != nil {
+				log.WithError(err).Error("Failed to get para heads")
+				return err
+			}
+			GetParaHeadData(header, parachainConn)
+		}
+	}
+}
+
+func GetParaheads(blockHash types.Hash, relaychainConn *relaychain.Connection) (string, error) {
+	type ParaId types.U32
+
+	type ParaHeadsQuery struct {
+		ParaId ParaId
+	}
+
+	query := ParaHeadsQuery{ParaId: 200}
+
+	encoded, err := types.EncodeToBytes(query)
+	if err != nil {
+		log.WithError(err).Error("Error")
+	}
+
+	allParaHeadsStorageKey, err := types.CreateStorageKey(
+		relaychainConn.GetMetadata(),
+		"Paras",
+		"Heads", encoded, nil)
+	if err != nil {
+		log.WithError(err).Error("Failed to create parachain header storage key")
+	}
+
+	// TODO1 - in polkadotjs webapp it seems possible to query this state via paras.heads without specifying any option
+	// and get all parachain heads as a response. eg: trying it on Rococo returns a lot of paraheads.
+	// I think we'll need this (to be able to check against the parachain headers hash
+	// of all headers in MMRLeaf.ParachainHeads), so we should get this working too.
+	response, err := relaychainConn.GetAPI().RPC.State.GetStorageRaw(allParaHeadsStorageKey, blockHash)
+	if err != nil {
+		log.WithError(err).Error("Failed to get parachain header")
+	}
+
+	// TODO2 - the above query returns some extra bytes, related the the HeadData type (try this state query in polkadotjs
+	// webapp for example). These extra bytes I think are the parachain ID, so the response type needs to account for
+	// this properly. the below is just a hack to get the actual header out. It's also not clear to me if the response
+	// contains the entire header, or just a hash of the header, or some truncated header? If it's the entire header,
+	// then great we can use it entirely instead of querying for it in a follow up call
```

Comment:
It should be the full parachain header, as you can see from this check: https://github.com/paritytech/polkadot/blob/master/node/core/candidate-validation/src/lib.rs#L418

Reply:
In order to decode that, we need the types of our parachain's header

Reply:
updated code with our parachain header decoding and it works, but it still requires removing the extra prefix bytes as in https://github.com/Snowfork/polkadot-ethereum/pull/378/files#diff-6bc10d8bbb796a09a8e3772904d4f0950335534c6a6b04d647c24fc66c04b6c8R139 - not sure what that prefix is for. maybe related to the Option<HeadData> type?

Reply:
if you query paraheads on rococo you can see that every header response, even from different chains, is prefixed with some bytes too in the response, though in the UI the prefix is removed

Reply:
@musnit those bytes are coming from the SCALE codec. The first byte signifies `Some` (`0x01`) or `None` (`0x00`), the next bytes are a compact encoded length prefix, in our case 4, but the prefix could have less bytes. 

Let me properly decode those later.